### PR TITLE
Tensor representation of weights

### PR DESCRIPTION
**Pull request by timmoon10**
Weights can now be represented as tensors. Since they are stored internally as distributed matrices, some tensor dimensions correspond to the matrix height and the rest to the matrix width. This should make it easier to export trained weights into a format readable by Tensorflow (#151). I am not sure if it will affect weight checkpointing.

Currently, none of the layers initialize weights using this new feature because I want to minimize merge conflicts with #171. It will just require simple modifications to the `setup_data` function in the base convolution, convolution, deconvolution, and batch normalization layers.

### CODE REVIEW

**Comment by ndryden** [comment]
This should be helpful for things like stacking convolutions channel-wise in a concatenation layer. 👍

File: include/lbann/weights/weights.hpp [hunk 1]
```diff
@@ -43,57 +41,79 @@ namespace lbann {
...
+                     std::vector<int> matrix_width_dims,
+                     El::Distribution col_dist,
+                     El::Distribution row_dist);
 
-  /** Get height of weights matrix. */
-  int get_height() const { return m_height; }
-  /** Get width of weights matrix. */
-  int get_width() const { return m_width; }
+  /** Get weight tensor dimensions. */
+  std::vector<int> get_dims() const;
```
Comment (samadejacobs) [inline include/lbann/weights/weights.hpp]: Could you please provide a note/example as part of documentation as to what each entry of the vector return  get_dims() (weight tensor dimensions) may represent e.g., dims[0] is N, dims[1] is channel etc?

Comment (timmoon10) [reply include/lbann/weights/weights.hpp]: I've expanded the documentation, although you should check if it is clear. Ultimately, this feature just implements a mathematical tensor and it is the responsibility of the layer class to interpret it.

**Comment by samadejacobs** [comment]
The setup function (line 128 in weights.cpp) that takes tensor dimensions as input parameter doesn't work, that is if I replace: 
this->m_weights[0]->setup(m_kernel_size / this->m_neuron_dims[0],
                              this->m_neuron_dims[0],
                              El::STAR, El::STAR);
with 
this->m_weights[0]->setup(this->m_kernel_dims)

The code should behave the same but it crashes with: 
Process 0 caught error message:
Nonconformal Gemm(T/C)N
Which I believe may have to do with not setting up  the matrix dimensions properly.

**Comment by timmoon10** [comment]
The convolution/deconvolution layers still expect the kernel matrix to have specific dimensions. When the internal implementation is changed, they should be more flexible.

As far as I am aware of, the only thing that's affected by the dimensions of the kernel matrix is quantization. Each matrix column is quantized separately. I think @ndryden had mixed results when quantizing convolution kernels, so this should be left as an expert mode feature.
